s3/src/main/java/ch/cyberduck/core/s3/S3Session.java

```diff
@@ -201,12 +206,29 @@ protected String getRestMetadataPrefix() {
     }
 
     @Override
-    protected RequestEntityRestStorageService connect(final Proxy proxy, final HostKeyCallback hostkey, final LoginCallback prompt, final CancelCallback cancel) {
+    protected RequestEntityRestStorageService connect(final Proxy proxy, final HostKeyCallback hostkey, final LoginCallback prompt, final CancelCallback cancel) throws BackgroundException {
         final HttpClientBuilder configuration = builder.build(proxy, this, prompt);
-        // Only for AWS
-        if(S3Session.isAwsHostname(host.getHostname())) {
-            configuration.setServiceUnavailableRetryStrategy(new S3TokenExpiredResponseInterceptor(this,
-                    new ThreadLocalHostnameDelegatingTrustManager(trust, host.getHostname()), key, prompt));
+        if(host.getProtocol().isOAuthConfigurable()) {
+            configuration.addInterceptorLast(oauth = new OAuth2RequestInterceptor(builder.build(ProxyFactory.get()
+                    .find(host.getProtocol().getOAuthAuthorizationUrl()), this, prompt).build(), host)
+                    .withRedirectUri(host.getProtocol().getOAuthRedirectUrl())
+                    .withFlowType(OAuth2AuthorizationService.FlowType.valueOf(host.getProtocol().getAuthorization())));
+            configuration.addInterceptorLast(sts = new STSAssumeRoleCredentialsRequestInterceptor(oauth, this, trust, key));
```

Comment:
As `sts` comes last, does this mean we renew `oauth` even if STS token is still valid? Is this necessary? Comment on this. Do we have tests for all possible cases?

Reply:
Implementation currently is intended as follows
1. Refresh of OAuth tokens when expired preemptively in `OAuth2RequestInterceptor` [^1] 
2. Refresh STS tokens using cached OAuth token preemptively in `STSAssumeRoleCredentialsRequestInterceptor` [^2]
3. On error response handle cases in `STSAssumeRoleTokenExpiredResponseInterceptor` as follows

* On `401`, `403` refresh OIDC Id token
* On `400` reuse cached OIDC Id token
* For both cases above refresh STS tokens

[^1]: https://github.com/iterate-ch/cyberduck/pull/14972/files#diff-7ec37d50b17b31a07e5438bce7f0f3f4eba8a930b1c100e12e3ad84dda666ff8R106
[^2]: https://github.com/iterate-ch/cyberduck/pull/14972/files#diff-c51064e4ed9ed159481473d881ec27aae4bd319978568b45480976f8f989b043R89

Reply:
Updated handling of authentication failures in `S3AuthenticationResponseInterceptor` as of 8ea744a1.

* Handle `403` and `400`
* Refresh STS token with cached OIDC Id token when not expired
* Otherwise refresh OIDC Id token first
* On failure refresh STS token with refreshed OIDC Id token

| Google OpenID Provider  | AWS API                       | HTTP Status Code | Error Code (in XML response body) |
|-------------------------|-------------------------------|------------------|-----------------------------------|
| OpenID Id Token Expired | S3 ListObjects                |                  | [^1]                              |
| OpenID Id Token Expired | STS AssumeRoleWithWebIdentity | 400              | `ExpiredTokenException` [^2]      |
| STS Tokens Expired      | S3 ListObjects                | 400              | `ExpiredToken` [^3]               |
| STS Tokens Invalid      | S3 ListObjects                | 400              | `InvalidToken` [^4]               |

[^1]:
[^2]: `Token expired: current date/time 1691790439 must be before the expiration date/time1691788167`

```
HTTP/1.1 400 Bad Request
x-amzn-RequestId: b6baa352-afcb-4763-a32c-a67d7eb5020b
Content-Type: text/xml
Content-Length: 346
Date: Sun, 13 Aug 2023 10:53:10 GMT
Connection: close

<ErrorResponse xmlns="https://sts.amazonaws.com/doc/2011-06-15/">
  <Error>
    <Type>Sender</Type>
    <Code>ExpiredTokenException</Code>
    <Message>Token expired: current date/time 1691923690 must be before the expiration date/time1691878565</Message>
  </Error>
  <RequestId>b6baa352-afcb-4763-a32c-a67d7eb5020b</RequestId>
</ErrorResponse>
```

[^3]: `The provided token has expired.`
[^3]: `The provided token is malformed or otherwise invalid.`